### PR TITLE
cluster train data

### LINKED ISSUE #311: API Guide 第2期：多机-准备分布式训练数据集（ 11月26日初稿pr，12月3日定稿merge）
# API Guide写前须知

## 背景
API Guide可以理解为是API的使用指南，旨在向用户说明API的设计思路、使用方法等
撰写计划共计约[45篇](http://agroup.baidu.com/paddlepaddle/view/office/1136638)。其中在PaddlePaddle v1.1中，已上线14篇，例如[官网-API-API使用指南-卷积](http://paddlepaddle.org/documentation/api/zh/1.1/api_guides/low_level/layers/conv.html)

## 任务说明
一篇完整的API Guide，需要包含以下几个内容：

1）对本篇主题的简介，例如：

``
卷积有两组输入：特征图和卷积核，依据输入特征和卷积核的形状、Layout不同、计算方式的不同，在Fluid里，有针对变长序列特征的一维卷积，有针对定长图像特征的二维(2D Conv)、三维卷积(3D Conv)，同时也有卷积计算的逆向过程，下面先介绍Fluid里的2D/3D卷积，再来介绍序列卷积。
``

2）对主题下每个API的介绍，例如

``
这个layer以一个mini batch的序列为输入，在每个序列内做softmax操作。其输出为一个mini batch相同shape的序列，但在序列内是经softmax归一化过的。 这个layer往往用于在每个sequence内做softmax归一化。
API Reference 请参考 sequence_softmax
``

- 完整例子：
[1.1已上线的API Guide](https://github.com/PaddlePaddle/FluidDoc/tree/release/1.1/doc/fluid/api/api_guides/low_level)

- 文档题目：请见标题

- 语言：汉语

- 格式：rst格式（因官网对rst格式支持较好，推荐用rst格式，如不了解rst，请在[这里](https://sphinx-doc-zh.readthedocs.io/en/latest/rest.html)学习）

- 字数：不限

- 效果预览：

为提高Review效率，请先在本地预览，确认格式正确再提交pr，请将预览效果截图贴在pr中，**不预览的pr不能合入**
预览工具的使用方法请见[这里](https://github.com/PaddlePaddle/FluidDoc#preview-with-ppo)

请注意：FluidDoc Repo的文档是通过Sphinx工具显示在官网上的。在Sphinx中，通过名为index.rst（或index_cn.rst等类似名称）的文件中的toctree来决定每一级显示哪几篇文档，在预览时，为了让文档能够正确预览，需要修改index，例如，如果新增一篇layers的API Guide，则需要修改`doc/fluid/api/api_guides/low_level/layers/index.rst` 这里的文件，并将自己文档的名字加入到index，如下图：

image

但在提交pr前，请不要git add 对index的修改，以避免多人同时修改这个文件引发conflict

- 提交分支：develop

- 提交位置：提交位置在`FluidDoc/doc/fluid/api/api_guides/low_level`

注：
layers文档，请提交到`FluidDoc/doc/fluid/api/api_guides/low_level/layers`
多机文档，请提交到`FluidDoc/doc/fluid/api/api_guides/low_level/cluster`

- pr提交截止日期：**请见标题**

- merge截止时间：**请见标题**

- 需要**特别注意**的格式问题

在一篇API Guide中，需要给出链接跳转至API Reference，[例如](http://paddlepaddle.org/documentation/api/zh/1.1/api_guides/low_level/layers/sequence.html#sequence-softmax)：

image

请注意：这里的超链接是用**内链**写成的，这样做的好处是，当某个API Reference的位置发生变化时，无需专门在文档里改超链接的路径，依然可以正确跳转。

用内链引用API Reference的方法： 以  ` :ref:<api_name>` 的格式引用reference（而不是其他格式：例如超链接、`xxx`等），`api_name`需与对应rst文件中的名字一致。

参考示例：https://raw.githubusercontent.com/PaddlePaddle/FluidDoc/develop_doc/doc/fluid/api_guides/low_level/optimizer/optimizer_all.rst

## 提交pr时的注意事项

- 贴预览效果

您可以选择以下两种方式展示预览效果：贴图或者给出预览网址，以方便我们检查格式

- 检查提交了几个文件

在预览工具时，需要修改index以确保文档可以显示在预览工具中，但在提交pr时，请去掉对index的修改，只提交您新写的文档，这是为了避免多人同时修改index文件引发conflict

- PR Reviewer：

为了提高Review效率，请在PR提交时增加Reviewer：一个PR的Reviewer由Peer Reviewer+ RD Reviewer+PM Reviewer组成，分别负责检查技术、格式的问题

Peer Reviewer ：1名，请找同方向的同学，例如NLP相关API请加kuke
RD Reviewer：luotao1
PM Reviewer：tink2123

- PR Labels：

为了方便管理API Guide进度，请在PR提交时增加Label `API Guide`

### PR DESCRIPTION
**Pull request by Yancey0623**
fixed #311 

预览链接：http://yq01-gpu-255-129-12-00.epc:8100/documentation/docs/zh/develop/api/api_guides/low_level/cluster/cluster_train_data_cn.html

### CODE REVIEW

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 1]
```diff
@@ -0,0 +1,108 @@
+..  _api_guide_cluster_train_data:
+
+####################
+分布式训练数据准备
+####################
+
+一个数据并行的分布式训练任务通常会含有多个训练节点，每个训练节点负责训练整个数据集种的一部分。所以在
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: 整个数据集种-》中，笔误

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+启动分布式训练任务之前需要将训练数据切分成多个小文件，通过一个 file_dispatcher 函数根据当前节点的
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: - 之前，需要将（断句更有利于阅读）
- file_dispatcher，这个函数，请使用内链
- “通过一个 file_dispatcher 函数根据当前节点的唯一序号(trainer_id)以及当前训练任务中训练节点的总数(trainers)决定读取哪一部分训练数据。”这句话太长了。

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+准备文本格式的分布式训练数据集
+------------------------------
+
+训练数据切分
+~~~~~~~~~~~~
+
+简单的，对于文本类训练数据来说，我们可以使用 split 命令将训练数据切分成多个小文件，例如：
+
+  .. code-block:: bash
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst:35]: 19/20行中间需加空行，目前预览中没有代码

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 4]
```diff
@@ -0,0 +1,108 @@
+..  _api_guide_cluster_train_data:
+
+####################
+分布式训练数据准备
+####################
+
+一个数据并行的分布式训练任务通常会含有多个训练节点，每个训练节点负责训练整个数据集种的一部分。所以在
+启动分布式训练任务之前需要将训练数据切分成多个小文件，通过一个 file_dispatcher 函数根据当前节点的
+唯一序号(trainer_id)以及当前训练任务中训练节点的总数(trainers)决定读取哪一部分训练数据。
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: 预览问题：
- 空格存在：请问是浏览器的问题，还是每行末尾断句问题。
image
比如“所以”和“在”中间有明显的空格。全文同。
- 预览和文档不匹配，比如预览中，没有file_dispatcher函数

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    cluster/housing.data.0003
+    cluster/housing.data.0004
+    cluster/housing.data.0000
+    cluster/housing.data.0001
+    cluster/housing.data.0005
+
+读取分布式训练数据集
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+在数据并行场景下，我们需要将训练数据平均分配给每个训练节点，通常的方法是实现一个函数，使之能够
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: 。（逗号改句号）通常的方法

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+根据当前任务的训练节点数量以及当前节点的唯一序号决定需要读取哪些文件，例如：
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: ，以及当前

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+  .. code-block:: python
+
+    def file_dispatcher(file_pattern, trainers, trainer_id):
+      file_list = glob.glob(file_pattern)
+      ret_list = []
+      for idx, f in enumerate(file_list):
+          if (idx + trainers) % trainers == trainer_id:
+              ret_list.append(f)
+      return ret_list
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: 预览中不是file_dispatcher，是gen_train_list

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+- file_pattern: 训练数据文件目录目录，上述例子可以是 `cluster/housing.data.*`
+- trainers: 当前任务的训练节点数。
+- trainer_id: 当前训练节点的唯一序号。
+
+准备 RecordIO 格式的分布式训练数据集
+-------------------------------------
+
+对于非文本类数据，可以预先将训练数据转换为 RecordIO 格式再进行训练, 并且转换成 RecordIO 格式
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: 。转换成 RecordIO 格式

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+的另一个好处是可以提升 IO 效率，从而提升分布式训练任务的运行效率。
+
+
+生成 RecordIO 格式数据集
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Fluid 提供了 `fluid.recordio_writer.convert_reader_to_recordio_files` API, 可以将训练数据转换成
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: fluid.recordio_writer.convert_reader_to_recordio_files 请使用内链

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  .. code-block:: bash
+
+      .
+      \_mnist-00000.recordio
+      |-mnist-00001.recordio
+      |-mnist-00002.recordio
+      |-mnist-00003.recordio
+      |-mnist-00004.recordio
+
+API Reference 请参考：:ref:`api_fluid_recordio_writer_convert_reader_to_recordio_file`
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: api_fluid_recordio_writer_convert_reader_to_recordio_file 这个单机方向已经不用了，并从最新官网撤下了：http://www.paddlepaddle.org/documentation/docs/zh/1.1/beginners_guide/index.html
请问多机还需要这个么？

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+读取 RecordIO 训练数据
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Fluid 种提供了 `fluid.layers.io.open_files` API 来读取 RecordIO 格式的训练数据，在以下样例代码
```
Comment (luotao1) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: - 种提供了-》中提供了
- `fluid.layers.io.open_files`，请使用内链

File: doc/fluid/api/api_guides/low_level/cluster/index.rst [hunk 12]
```diff
@@ -0,0 +1,9 @@
+==========
+多机训练
+==========
+
+..  toctree::
+    :maxdepth: 1
```
Comment (shanyi15) [inline doc/fluid/api/api_guides/low_level/cluster/index.rst]: 为了避免多人同时修改index造成冲突，请取消对这个文件的修改

File: doc/fluid/api/api_guides/index.rst [hunk 13]
```diff
@@ -7,6 +7,7 @@ API使用指南
 
     high_low_level_api.md
     low_level/layers/index.rst
+    low_level/cluster/index.rst
```
Comment (shanyi15) [inline doc/fluid/api/api_guides/index.rst]: 为了避免多人同时修改index文件造成冲突，请取消对这个文件的修改，最后由我们统一来加～谢谢

File: .gitignore [hunk 14]
```diff
@@ -0,0 +1 @@
+.vscode/
```
Comment (shanyi15) [inline .gitignore:1]: 请问这个文件是否是添加错了呢

Comment (Yancey0623) [reply .gitignore:1]: 这个没有，如果是用vscode作为IDE，会自动在工程目录下生成`.vscode/`目录，加在gitignore里比较方便。 Paddle里也有：https://github.com/PaddlePaddle/Paddle/blob/develop/.gitignore#L15

File: doc/fluid/api/api_guides/low_level/cluster/index.rst (hunk 12 above)
Comment (Yancey0623) [reply doc/fluid/api/api_guides/low_level/cluster/index.rst]: done.

File: doc/fluid/api/api_guides/index.rst (hunk 13 above)
Comment (Yancey0623) [reply doc/fluid/api/api_guides/index.rst]: done.

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst (hunk 10 above)
Comment (Yancey0623) [reply doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst]: 根据hi群里的讨论，已去掉此demo。

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst [hunk 15]
```diff
@@ -0,0 +1,64 @@
+..  _api_guide_cluster_train_data:
+
+####################
+分布式训练数据准备
```
Comment (typhoonzero) [inline doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst:4]: 分布式训练reader准备

File: doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst (hunk 3 above)
Comment (typhoonzero) [reply doc/fluid/api/api_guides/low_level/cluster/cluster_train_data_cn.rst:35]: 还是缺空格？

**Comment by luotao1** [comment]
LGTM!
@shanyi15 @tink2123 可以merge后，再帮忙修改下https://github.com/PaddlePaddle/FluidDoc/pull/393#discussion_r236972527么？

**Review by shanyi15 — APPROVED** [review]
LGTM
